keps/sig-cluster-lifecycle/0029-20180918-kubeadm-phases-beta.md

```diff
@@ -0,0 +1,184 @@
+  interface.
+- To significantly reduces the effort for maintaining `phases` up to date
+  with `kubeadm init`.
+- To support extension of the "phases" concept to other kubeadm workflows.
+- To enable re-use of phases across different workflows e.g. the cert phase
+  used by the `kubeadm init` and by the `kubeadm join --master` workflows.
+
+### Non-goals
+
+- This proposal doesn't include any changes of improvements to the actual `kubeadm init`
+  workflow.
+- This proposal doesn't include implementation of workflows different than `kubeadm init`;
+  nevertheless, this proposal introduces a framework that will allow such implementation in future.
+
+## Proposal
+
+### User Stories
+
+- As a kubernetes administrator/IT automation tool, I want to run all the phases of
+  the `kubeadm init` workflow.
+- As a kubernetes administrator/IT automation tool, I want to run only one or some phases
+  of the `kubeadm init` workflow.
+- As a kubernetes administrator/IT automation tool, I want to run all the phases of
+  the `kubeadm init` workflow except some phases.
+
+### Implementation Details
+
+The core of the new phase design consist into a simple, lightweight workflow manager to be used
+for implementing composable kubeadm workflows.
+
+Composable kubeadm workflows are build by an ordered sequence of phases; each phase can have it's
+own, nested, ordered sequence of phases. For instance:
+
+```bash
+  preflight       Run master pre-flight checks
+  certs           Generates all PKI assets necessary to establish the control plane
+    /ca             Generates a self-signed kubernetes CA 
+    /apiserver      Generates an API server serving certificate and key
+    ...
+  kubeconfig      Generates all kubeconfig files necessary to establish the control plane 
+    /admin          Generates a kubeconfig file for the admin to use and for kubeadm itself
+    /kubelet        Generates a kubeconfig file for the kubelet to use.
+    ...
+  ...
+````
+
+Phases implementation will be designed in order to ensure that phases are reusable across different
+workflows thus allowing e.g. reuse of phase `certs` in both `kubeadm init` and
+`kubeadm join --master` workflows.
+
+Each kubeadm workflow will be defined and managed using an `Executor` that basically is a component
+in charge of of running phases according to the given order; nested phases will be executed
+immediately after their parent phase.
+
+The `Executor` will be instantiated by kubeadm commands, and the executor options exposed as
+kubeadm command flags thus allowing to filter the list of phases to be executed. As a consequence
+the `kubeadm init` UX will be modified from:
+
+```bash
+Usage:
+  kubeadm init [flags]
+```
+
+to
+
+```bash
+Usage:
+  kubeadm init [workflow-flags] [flags]
+```
+
+[workflow-flags] will be a set of optional flags that allows user to select which phase
+of the `kubeadm init` logic to execute (`--only-phases`) or to skip (`--skip-phases`). e.g.
```

Comment:
my vote goes to something in the lines of:
`--apply-phases`

i.e. having a verb in the flag name.

Reply:
@neolit123, thanks for this comments! They add more value to this proposal!

> we need to outline ordering to the user somehow

Yes, definitely. I'm going to add a sentence about this

> if --apply-phases=addons, kubelet is called,  would that imply that the user wants to first install addons and then run the kubelet

We can choose if to allow execution of a single phase only - as of today - or if to allow execution of one or many phases - as it is now in this proposal -. If we confirm many, IMO the executor should always respect the given order of phases for sake of consistency and for reducing complexity (only one behaviour to be implemented and tested).

Eventually we should consider if to give evidence of that to the user by printing the execution plan - the ordered list of phases that are going to be executed - when running in --dry-run  mode

> have you considered  `kubeadm init --phase-config=<some-config-structure>`

This is discussed in the drawbacks paragraph, and the answer to your question is yes, this was considered even if not in this exact form, but I discarded this option for several reasons, the most important are again consistency and the ambition to reduce complexity.

Phases are part of kubeadm init workflow, so flags of phases are expected to be a subset of the flags of the kubeadm workflow; if a flag is added to kubeadm init workflow, phases should get the same flag, and I think that the viceversa holds as well.

On the contrary, If we open to something different there will be again flag proliferation, and both the CLI help, the user documentation and the code for flag validation as well will be forced to manage an O(MxN) problem.

Reply:
> We should have strict dependency ordering and go on that lines

Totally agree. See "IMO the executor should always respect the given order of phases for sake of consistency and for reducing complexity (only one behaviour to be implemented and tested)" above.

> This is no longer a full blown init, but rather kubeadm phases sub-command.

I got your point but I'm not sure I agree with you.
1) what will happen if we apply the phases concept to other kubeadm commands, like kubeadm upgrade or kubeadm join? Everything (all the phases for all the commands) will clash into kubeadm phases? 
2) having phases separated by the workflow they belong already created problems in the past, and today there is set of utility commands under phases which are not technically a phases of the kubeadm init workflow e.g certs renew. I want to prevent this happening again otherwise the user won`t understand anymore what a phase is.
3) executing a phase is nevertheless executing a piece of init , so why to move it in a different place?

Reply:
As agree in kubeadm office hours we are giving the users only the capability to invoke one phase at time